core/src/Cuda/Kokkos_CudaSpace.cpp

```diff
@@ -434,49 +583,30 @@ void SharedAllocationRecord<Kokkos::CudaHostPinnedSpace, void>::deallocate(
 // <editor-fold desc="SharedAllocationRecord destructors"> {{{1
 
 SharedAllocationRecord<Kokkos::CudaSpace, void>::~SharedAllocationRecord() {
-#if defined(KOKKOS_ENABLE_PROFILING)
-  if (Kokkos::Profiling::profileLibraryLoaded()) {
-    SharedAllocationHeader header;
-    Kokkos::Impl::DeepCopy<CudaSpace, HostSpace>(
-        &header, RecordBase::m_alloc_ptr, sizeof(SharedAllocationHeader));
-
-    Kokkos::Profiling::deallocateData(
-        Kokkos::Profiling::make_space_handle(Kokkos::CudaSpace::name()),
-        header.m_label, data(), size());
-  }
-#endif
-
-  m_space.deallocate(SharedAllocationRecord<void, void>::m_alloc_ptr,
-                     SharedAllocationRecord<void, void>::m_alloc_size);
+  SharedAllocationHeader header;
+  Kokkos::Impl::DeepCopy<Kokkos::CudaSpace, HostSpace>(
+      &header, RecordBase::m_alloc_ptr, sizeof(SharedAllocationHeader));
+  auto alloc_size = SharedAllocationRecord<void, void>::m_alloc_size;
+  m_space.deallocate(header.label(),
+                     SharedAllocationRecord<void, void>::m_alloc_ptr,
+                     alloc_size, (alloc_size - sizeof(SharedAllocationHeader)));
 }
 
 SharedAllocationRecord<Kokkos::CudaUVMSpace, void>::~SharedAllocationRecord() {
-#if defined(KOKKOS_ENABLE_PROFILING)
-  if (Kokkos::Profiling::profileLibraryLoaded()) {
-    Cuda::impl_static_fence();  // Make sure I can access the label ...
-    Kokkos::Profiling::deallocateData(
-        Kokkos::Profiling::make_space_handle(Kokkos::CudaUVMSpace::name()),
-        RecordBase::m_alloc_ptr->m_label, data(), size());
-  }
-#endif
-
-  m_space.deallocate(SharedAllocationRecord<void, void>::m_alloc_ptr,
-                     SharedAllocationRecord<void, void>::m_alloc_size);
+  m_space.deallocate(RecordBase::m_alloc_ptr->m_label,
```

Comment:
ah we need to put a temporary char* here as the label, and only fill the ptr with `m_alloc_ptr->m_label` if `profileLibraryLoaded()` is true. Otherwise we add an unnecessary UM page migration here.

Reply:
Addressed